exercises/saddle-points/src/example.c

```diff
@@ -1,73 +1,45 @@
 #include "saddle_points.h"
+
 #include <stdlib.h>
 #include <stdbool.h>
 
-static saddle_points_t *noSaddlePoints(void)
+static bool is_saddle_point(size_t nrows,
+                            size_t ncols,
+                            size_t row,
+                            size_t col, uint8_t matrix[nrows][ncols])
 {
-   saddle_points_t *saddle_points = malloc(sizeof(saddle_points_t));
-
-   if (saddle_points)
-      saddle_points->count = 0;
 
-   return saddle_points;
-}
-
-static saddle_points_t *add(saddle_points_t * saddle_points, size_t row,
-                            size_t column)
-{
-   size_t count = 0;
-   if (saddle_points)
-      count = saddle_points->count;
-
-   saddle_points_t *new_saddle_points = realloc(saddle_points,
-                                                sizeof(saddle_points_t) +
-                                                sizeof(saddle_point_t) *
-                                                ++count);
-
-   if (new_saddle_points) {
-      saddle_points = new_saddle_points;
-      saddle_points->count = count;
-      saddle_points->points[count - 1] = (saddle_point_t) {
-      row, column};
-   } else {
-      free(saddle_points);
-      saddle_points = NULL;
+   for (size_t i = 0; i < nrows; i++) {
+      if (matrix[row][col] > matrix[i][col]) {
+         return false;
+      }
    }
 
-   return saddle_points;
-}
-
-static bool is_greatest_in_row(int16_t row, int16_t column, size_t columns,
-                               uint8_t matrix[][columns])
-{
-   uint8_t candidate = matrix[row][column];
-   for (size_t c = 0; c < columns; ++c)
-      if (matrix[row][c] > candidate)
+   for (size_t i = 0; i < ncols; i++) {
+      if (matrix[row][col] < matrix[row][i]) {
          return false;
-   return true;
-}
+      }
+   }
 
-static bool is_least_in_column(int16_t row, int16_t column, size_t rows,
-                               size_t columns, uint8_t matrix[][columns])
-{
-   uint8_t candidate = matrix[row][column];
-   for (size_t r = 0; r < rows; ++r)
-      if (matrix[r][column] < candidate)
-         return false;
    return true;
 }
 
-saddle_points_t *saddlePoints(size_t rows, size_t columns,
-                              uint8_t matrix[][columns])
+saddle_points_t *saddle_points(size_t nrows,
+                               size_t ncols, uint8_t matrix[nrows][ncols])
 {
-   saddle_points_t *saddle_points = noSaddlePoints();
 
-   if (rows > 0 && columns > 0 && matrix != NULL)
-      for (size_t row = 0; row < rows; ++row)
-         for (size_t column = 0; column < columns; ++column)
-            if (is_greatest_in_row(row, column, columns, matrix)
-                && is_least_in_column(row, column, rows, columns, matrix))
-               saddle_points = add(saddle_points, row + 1, column + 1);
+   saddle_points_t *saddle = calloc(1, sizeof(saddle_points_t));
+   saddle->points = calloc(nrows * ncols, sizeof(saddle_point_t));
```

Comment:
doesn't this allocate too many items? `rows * columns` must be significantly more than the number of actual saddle points in most cases.

Reply:
I don't think we have a way to enforce this as a requirement in the test's though.
Interesting question: should the example here show best practice, or simply adhere to the tests?
More pertinent question: is there a way to do this that removes the extra allocated memory yet is still accessible to students?

Reply:
IMO we should just move the `free` internal to this module and expose it via a public function. Then we can have a correct solution that doesn't impose any extra requirements on submitted solutions. It would also be a relatively minor change to the existing solution.

Reply:
do you mean to provide a function in header file, so students implement their own free function ?
```
void free_saddle_points(saddle_points_t saddle_points);
```

Reply:
@gabriel376 yes. That way the implementation owns allocating and freeing.

Reply:
great!
I will update my PR to implement that
thank you!